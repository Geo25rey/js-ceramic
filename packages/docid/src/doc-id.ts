import CID from 'cids';
import multibase from 'multibase';
import doctypes from './doctype-table';
import varint from 'varint';
import uint8ArrayConcat from 'uint8arrays/concat';
import uint8ArrayToString from 'uint8arrays/to-string';
import { DEFAULT_BASE, DOCID_CODEC } from './constants';

const getKey = (obj: { [key: string]: number }, value: number): string | undefined => {
  for (const [k, v] of Object.entries(obj)) {
    if (v === value) return k;
  }
};

// Definition
// '<multibase-prefix><multicodec-docid><doctype><genesis-cid-bytes>'
// '<multibase-prefix><multicodec-docid><doctype><genesis-cid-bytes><commit-cid-bytes>'

export class DocID {
  /**
   * Create a new DocID.
   *
   * @param {string|number}      doctype
   * @param {CID|string}         cid
   * @param {CID|string}         commit CID. Pass '0' as shorthand for the genesis commit.
   * @param {string}             [multibaseName = 'base36']
   *
   * @example
   * new DocID(<docType>, <CID>|<cidStr>)
   * new DocID(<docType>, <CID>|<cidStr>, <CommitCID>|<CommitCidStr>, <multibaseName>)
   */

  readonly #doctype: number;
  readonly #cid: CID;
  private _bytes: Uint8Array;

  constructor(doctype: string | number, cid: CID | string) {
    if (!cid) throw new Error('constructor: cid required');
    this.#doctype = typeof doctype === 'string' ? doctypes[doctype] : doctype;
    if (!this.#doctype && this.#doctype !== 0) throw new Error('constructor: doctype required');
    this.#cid = typeof cid === 'string' ? new CID(cid) : cid;
  }

  /**
   * Copies the given DocID and returns a copy of it, optionally changing the commit to the one provided
   * @param other
   */
  static fromOther(other: DocID): DocID {
    return new DocID(other.type, other.cid);
  }

  static fromBytes(bytes: Uint8Array): DocID {
    const docCodec = varint.decode(bytes);
    if (docCodec !== DOCID_CODEC) throw new Error('fromBytes: invalid docid, does not include docid codec');
    bytes = bytes.slice(varint.decode.bytes);
    const docType = varint.decode(bytes);
    bytes = bytes.slice(varint.decode.bytes);

    let cid;

    try {
      cid = new CID(bytes);
    } catch (e) {
      // Includes commit
      const cidLength = DocID._genesisCIDLength(bytes);
      cid = new CID(bytes.slice(0, cidLength));
    }

    return new DocID(docType, cid);
  }

  static _genesisCIDLength(bytes: Uint8Array): number {
    let offset = 0;

    varint.decode(bytes); // cid version
    offset += varint.decode.bytes;

    varint.decode(bytes.slice(offset)); // cid codec
    offset += varint.decode.bytes;

    varint.decode(bytes.slice(offset)); //mh codec
    offset += varint.decode.bytes;

    const length = varint.decode(bytes.slice(offset)); //mh length
    return offset + length + 1;
  }

  static fromString(docId: string): DocID {
    docId = docId.split('ceramic://').pop();
    // Likely temp, remove legacy once all ceramic update, but should make updating easier
    docId = docId.split('/ceramic/').pop();
    if (docId.includes('commit')) {
      docId = docId.split('?')[0];
    }
    const bytes = multibase.decode(docId);
    return DocID.fromBytes(bytes);
  }

  /**
   * Get doc type code
   *
   * @returns {number}
   * @readonly
   */
  get type(): number {
    return this.#doctype;
  }

  /**
   * Get doc type name
   *
   * @returns {string}
   * @readonly
   */
  get typeName(): string {
    const name = getKey(doctypes, this.#doctype);
    if (!name) throw new Error('docTypeName: no registered name available');
    return name;
  }

  /**
   * Get CID
   *
   * @returns {CID}
   * @readonly
   */
  get cid(): CID {
    return this.#cid;
  }

  /**
   * Get CID codec name string
   *
   * @returns {string}
   * @readonly
   */
  get codec(): string {
    return this.#cid.codec;
  }

  /**
   * Get bytes of DocId
   *
   * @returns {Uint8Array}
   * @readonly
   */
  get bytes(): Uint8Array {
    if (this._bytes == null) {
      const codec = varint.encode(DOCID_CODEC);
      const doctype = varint.encode(this.type);

      this._bytes = uint8ArrayConcat([codec, doctype, this.cid.bytes]);
    }
    return this._bytes;
  }

  /**
   * Compare equality with another DocID.
   *
   * @param   {DocID}   other
   * @returns {bool}
   */
  equals(other: DocID | Uint8Array | string): boolean {
    let otherDocID;
    if (typeof other === 'string') otherDocID = DocID.fromString(other);
    else if (other instanceof Uint8Array) {
      otherDocID = DocID.fromBytes(other);
    } else {
      otherDocID = other;
    }

    return this.type === otherDocID.type && this.cid.equals(otherDocID.cid);
  }

  /**
   * Encode the DocID into a string.
   */
  toString(): string {
    return uint8ArrayToString(multibase.encode(DEFAULT_BASE, this.bytes));
  }

  /**
   * Encode the DocID into a base36 url
   *
   * @returns {string}
   */
  toUrl(): string {
    return `ceramic://${this.toString()}`;
  }

  /**
   * DocId(k3y52l7mkcvtg023bt9txegccxe1bah8os3naw5asin3baf3l3t54atn0cuy98yws)
   *
   * @returns {String}
   */
  [Symbol.for('nodejs.util.inspect.custom')](): string {
    return `DocID(${this.toString()})`;
  }

  [Symbol.toPrimitive](): string | Uint8Array {
    return this.toString();
  }

  /**
   * Determine if given DocID, DocID string or bytes is a valid DocID
   *
   * @param   {any}     other
   * @returns {Boolean}
   */
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  static isDocID(other: any): boolean {
    try {
      if (typeof other === 'string') {
        DocID.fromString(other);
        return true;
      } else if (other instanceof Uint8Array) {
        DocID.fromBytes(other);
        return true;
      } else {
        return (other.type || other.type === 0) && Boolean(other.cid);
      }
    } catch (e) {
      return false;
    }
  }
}
