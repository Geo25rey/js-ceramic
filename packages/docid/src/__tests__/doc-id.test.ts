import { DocID } from '../doc-id';
import CID from 'cids'

const cidStr = 'bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a'
const docIdStr = 'kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'
const docIdUrl = 'ceramic://kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'
const docIdLegacy = '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'

const docIdStrCommit = 'k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju'
const docIdStrCommit0 = 'k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4'
const docIdLegacyCommit = '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s?commit=bagjqcgzaday6dzalvmy5ady2m5a5legq5zrbsnlxfc2bfxej532ds7htpova'
const docIdLegacyCommit0 = '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s?commit=0'

const docIdBytes = new Uint8Array([
  206,   1,   0,   1, 133,   1,  18,  32,  84,
179, 109,  86,  78, 173, 242, 220,  63, 166,
 43,  94,  52, 240,  80, 102, 174, 138, 255,
216,  30, 185, 101,  36, 150,  57, 240, 167,
 85, 209, 179, 188
])

const docIdBytesCommit = new Uint8Array([
  206,   1,   0,   1, 133,   1,  18,  32,  84, 179, 109,  86,
   78, 173, 242, 220,  63, 166,  43,  94,  52, 240,  80, 102,
  174, 138, 255, 216,  30, 185, 101,  36, 150,  57, 240, 167,
   85, 209, 179, 188,   1, 147,   1,  27,  32,  24,  49, 225,
  228,  11, 171,  49, 208,  15,  26, 103,  65, 213, 144, 208,
  238,  98,  25,  53, 119,  40, 180,  18, 220, 137, 238, 244,
   57, 124, 243, 123, 170
])

const docIdBytesCommit0 = new Uint8Array([
  206,   1,   0,   1, 133,   1,  18,  32,  84,
  179, 109,  86,  78, 173, 242, 220,  63, 166,
   43,  94,  52, 240,  80, 102, 174, 138, 255,
  216,  30, 185, 101,  36, 150,  57, 240, 167,
   85, 209, 179, 188,   0
])

describe('DocID', () => {

  it('create by parts (type:int, cid:cid)',  () => {
    const type = 0
    const cid = new CID(cidStr)
    const docid = new DocID(type, cid)

    expect(docid.type).toEqual(type)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from bytes ',  () => {
    const docid = DocID.fromBytes(docIdBytes)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from bytes inlcuding commit',  () => {
    const docid = DocID.fromBytes(docIdBytesCommit)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from bytes inlcuding commit 0',  () => {
    const docid = DocID.fromBytes(docIdBytesCommit0)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from string',  () => {
    const docid = DocID.fromString(docIdStr)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from string including commit',  () => {
    const docid = DocID.fromString(docIdStrCommit)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from string including commit 0',  () => {
    const docid = DocID.fromString(docIdStrCommit0)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from url string',  () => {
    const docid = DocID.fromString(docIdUrl)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from legacy string "/ceramic/"',  () => {
    const docid = DocID.fromString(docIdLegacy)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from legacy string "/ceramic/" with commit param "?commit="',  () => {
    const docid = DocID.fromString(docIdLegacyCommit)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('create from legacy string "/ceramic/" with commit param "?commit=0"',  () => {
    const docid = DocID.fromString(docIdLegacyCommit0)

    expect(docid.type).toEqual(0)
    expect(docid.cid.toString()).toEqual(cidStr)
    expect(docid.toString()).toMatchSnapshot()
  })

  it('roundtrip docID string',  () => {
    const docid = new DocID('tile', cidStr)
    const docid2 = DocID.fromString(docid.toString())
    expect(docid.toString()).toEqual(docid2.toString())
  })

  it('roundtrip docID bytes',  () => {
    const docid = new DocID('tile', cidStr)
    const docid2 = DocID.fromBytes(docid.bytes)
    expect(docid.toString()).toEqual(docid2.toString())
  })

  it('.bytes',  () => {
    const docid = new DocID('tile', cidStr)
    const bytes = docid.bytes
    expect(bytes).toBeDefined()
    expect(bytes instanceof Uint8Array).toEqual(true)
    expect(bytes).toMatchSnapshot()
  })

  it('.typeName if registered',  () => {
    const docid = new DocID('tile', cidStr)
    expect(docid.typeName).toEqual('tile')
    const docid2 = new DocID(10, cidStr)
    expect(() => docid2.typeName).toThrowErrorMatchingSnapshot()
  })

  it('.toString()',  () => {
    const docid = new DocID('tile', cidStr)
    const str = docid.toString()
    expect(str).toBeDefined()
    expect(str).toMatchSnapshot()
  })

  it('.toUrl()',  () => {
    const docid = new DocID('tile', cidStr)
    const str = docid.toUrl()
    expect(str).toBeDefined()
    expect(str.includes('ceramic://')).toEqual(true)
    expect(str).toMatchSnapshot()
  })

  it('equals other DocID',  () => {
    const docid = new DocID('tile', cidStr)
    const docid2 = new DocID('tile', cidStr)
    const docid3 = new DocID('caip10-link', cidStr)
    expect(docid.equals(docid2)).toEqual(true)
    expect(docid.equals(docid3)).toEqual(false)
  })
})
