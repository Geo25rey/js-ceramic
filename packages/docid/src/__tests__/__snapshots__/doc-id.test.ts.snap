// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocID .bytes 1`] = `
Uint8Array [
  206,
  1,
  0,
  1,
  133,
  1,
  18,
  32,
  84,
  179,
  109,
  86,
  78,
  173,
  242,
  220,
  63,
  166,
  43,
  94,
  52,
  240,
  80,
  102,
  174,
  138,
  255,
  216,
  30,
  185,
  101,
  36,
  150,
  57,
  240,
  167,
  85,
  209,
  179,
  188,
]
`;

exports[`DocID .toString() 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID .toUrl() 1`] = `"ceramic://kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID .typeName if registered 1`] = `"No doctype registered for index 10"`;

exports[`DocID create by parts (type:int, cid:cid) 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from bytes  1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from bytes inlcuding commit 0 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from bytes inlcuding commit 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from legacy string "/ceramic/" 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from legacy string "/ceramic/" with commit param "?commit=" 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from legacy string "/ceramic/" with commit param "?commit=0" 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from string 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from string including commit 0 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from string including commit 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`DocID create from url string 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`nodejs inspect 1`] = `"DocID(kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s)"`;
